PR: https://github.com/odoo/odoo/pull/164671

From: 505a5ca54f8f4e204352b69d4388786c20933181
From: std-odoo
Date: 2024-09-06 09:45:55

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'ir.qweb'

Total Changes: 110

[IMP] mail: allow non-template editors users to use simple expressions

Purpose
=======
Allow the non-template editor users to write simple expression
(e.g. object.name...) in mass-mailing but also in mail templates.

The mail composer in mass-mode is still in readonly mode if we are not
template editor, and we selected a template, because it might contain
complex code (conditions, loops, etc).

Technical
=========
The default value for qweb can be done either with
```
<t t-out="name or 'default'"/>
```

or with
```
<t t-out="name">default</t>
```

But, inline qweb only support this syntax for now
```
{{name or 'default'}}
```

making the security check more complex, since we need to parse the code

This commit add a new syntax for the default value of inline qweb,
which will make the security check trivial.

Task-3877116

Part-of: odoo/odoo#164671
Related: odoo/enterprise#69118
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/mail/models/ir_qweb.py
+++ b/addons/mail/models/ir_qweb.py
@@ -1,16 +1,60 @@
 # -*- coding: utf-8 -*-
-from odoo import models
+
+from odoo import api, models
+
 
 class IrQweb(models.AbstractModel):
-    """ Add ``raise_on_code`` option for qweb. When this option is activated
-    then all directives are prohibited.
+    """Add ``raise_on_forbidden_code`` option for qweb.
+
+    When this option is activated, only a whitelist of expressions is allowed.
     """
-    _inherit = 'ir.qweb'
+
+    _inherit = "ir.qweb"
+
+    allowed_directives = (
+        "esc",
+        "out",
+        "inner-content",
+        "att",
+        "tag-open",
+        "tag-close",
+    )
 
     def _get_template_cache_keys(self):
-        return super()._get_template_cache_keys() + ['raise_on_code']
+        return super()._get_template_cache_keys() + ["raise_on_forbidden_code"]
+
+    def _compile_directive(self, el, compile_context, directive, level):
+        if (
+            compile_context.get("raise_on_forbidden_code")
+            and directive not in self.allowed_directives
+        ):
+            raise PermissionError("This directive is not allowed for this rendering mode.")
+        return super()._compile_directive(el, compile_context, directive, level)
+
+    def _compile_directive_att(self, el, compile_context, level):
+        if compile_context.get("raise_on_forbidden_code"):
+            if set(el.attrib) - {"t-esc", "t-out", "t-tag-open", "t-tag-close"}:
+                raise PermissionError("This directive is not allowed for this rendering mode.")
+
+        return super()._compile_directive_att(el, compile_context, level)
+
+    def _compile_expr(self, expr, raise_on_missing=False):
+        if self.env.context.get("raise_on_forbidden_code") and not self._is_expression_allowed(expr):
+            raise PermissionError("This directive is not allowed for this rendering mode.")
+        return super()._compile_expr(expr, raise_on_missing)
+
+    def _is_expression_allowed(self, expression):
+        return expression.strip() in self.allowed_qweb_expressions()
 
-    def _compile_directives(self, el, compile_context, indent):
-        if compile_context.get('raise_on_code'):
-            raise PermissionError("This rendering mode prohibits the use of directives.")
-        return super()._compile_directives(el, compile_context, indent)
+    @api.model
+    def allowed_qweb_expressions(self):
+        # QWeb expressions allowed if we are not template editor
+        return (
+            "object.name",
+            "object.contact_name",
+            "object.partner_id",
+            "object.partner_id.name",
+            "object.user_id",
+            "object.user_id.name",
+            "object.user_id.signature",
+        )

--- a/addons/mail/models/mail_render_mixin.py
+++ b/addons/mail/models/mail_render_mixin.py
@@ -76,7 +76,7 @@ class MailRenderMixin(models.AbstractModel):
             if sub_field_name:
                 expression += "." + sub_field_name
             if null_value:
-                expression += " or '''%s'''" % null_value
+                expression += f" ||| {null_value}"
             expression += " }}"
         return expression
 
@@ -205,24 +205,24 @@ class MailRenderMixin(models.AbstractModel):
     # SECURITY
     # ------------------------------------------------------------
 
-    def _is_dynamic(self):
+    def _has_unsafe_expression(self):
         for template in self.sudo():
             for fname, field in template._fields.items():
                 engine = getattr(field, 'render_engine', 'inline_template')
                 if engine in ('qweb', 'qweb_view'):
-                    if self._is_dynamic_template_qweb(template[fname]):
+                    if self._has_unsafe_expression_template_qweb(template[fname]):
                         return True
                 else:
-                    if self._is_dynamic_template_inline_template(template[fname]):
+                    if self._has_unsafe_expression_template_inline_template(template[fname]):
                         return True
         return False
 
     @api.model
-    def _is_dynamic_template_qweb(self, template_src):
+    def _has_unsafe_expression_template_qweb(self, template_src):
         if template_src:
             try:
                 node = html.fragment_fromstring(template_src, create_parent='div')
-                self.env["ir.qweb"].with_context(raise_on_code=True)._compile(node)
+                self.env["ir.qweb"].with_context(raise_on_forbidden_code=True)._compile(node)
             except QWebException as e:
                 if isinstance(e.__cause__, PermissionError):
                     return True
@@ -230,18 +230,19 @@ class MailRenderMixin(models.AbstractModel):
         return False
 
     @api.model
-    def _is_dynamic_template_inline_template(self, template_txt):
+    def _has_unsafe_expression_template_inline_template(self, template_txt):
         if template_txt:
             template_instructions = parse_inline_template(str(template_txt))
-            if len(template_instructions) > 1 or template_instructions[0][1]:
+            expressions = [inst[1] for inst in template_instructions]
+            if not all(self.env["ir.qweb"]._is_expression_allowed(e) for e in expressions if e):
                 return True
         return False
 
     def _check_access_right_dynamic_template(self):
-        if not self.env.su and not self.env.user.has_group('mail.group_mail_template_editor') and self._is_dynamic():
+        if not self.env.su and not self.env.user.has_group('mail.group_mail_template_editor') and self._has_unsafe_expression():
             group = self.env.ref('mail.group_mail_template_editor')
             raise AccessError(
-                _('Only users belonging to the "%(group_name)s" group can modify dynamic templates.',
+                _('Only members of %(group_name)s group are allowed to edit templates containing sensible placeholders',
                   group_name=group.name)
             )
 
@@ -309,7 +310,7 @@ class MailRenderMixin(models.AbstractModel):
                 render_result = self.env['ir.qweb']._render(
                     html.fragment_fromstring(template_src, create_parent='div'),
                     variables,
-                    raise_on_code=is_restricted,
+                    raise_on_forbidden_code=is_restricted,
                     **(options or {})
                 )
                 # remove the rendered tag <div> that was added in order to wrap potentially multiples nodes into one.
@@ -318,7 +319,7 @@ class MailRenderMixin(models.AbstractModel):
                 if isinstance(e, QWebException) and isinstance(e.__cause__, PermissionError):
                     group = self.env.ref('mail.group_mail_template_editor')
                     raise AccessError(
-                        _('Only users belonging to the "%(group_name)s" group can modify dynamic templates.',
+                        _('Only members of %(group_name)s group are allowed to edit templates containing sensible placeholders',
                            group_name=group.name)
                     ) from e
                 _logger.info("Failed to render template: %s", template_src, exc_info=True)
@@ -405,29 +406,20 @@ class MailRenderMixin(models.AbstractModel):
 
         :return dict: {res_id: string of rendered template based on record}
         """
-        results = dict.fromkeys(res_ids, u"")
+        results = dict.fromkeys(res_ids, "")
         if not template_txt or not res_ids:
             return results
 
-        template_instructions = parse_inline_template(str(template_txt))
-        is_dynamic = len(template_instructions) > 1 or template_instructions[0][1]
-
-        if (not self._unrestricted_rendering and is_dynamic and not self.env.is_admin() and
-           not self.env.user.has_group('mail.group_mail_template_editor')):
+        if (not self._unrestricted_rendering
+            and not self.env.is_admin()
+            and not self.env.user.has_group('mail.group_mail_template_editor')
+            and self._has_unsafe_expression_template_inline_template(str(template_txt))):
             group = self.env.ref('mail.group_mail_template_editor')
             raise AccessError(
-                _('Only users belonging to the "%(group_name)s" group can modify dynamic templates.',
+                _('Only members of %(group_name)s group are allowed to edit templates containing sensible placeholders',
                   group_name=group.name)
             )
 
-        if not is_dynamic:
-            # Either the content is a raw text without placeholders, either we fail to
-            # detect placeholders code. In both case we skip the rendering and return
-            # the raw content, so even if we failed to detect dynamic code,
-            # non "mail_template_editor" users will not gain rendering tools available
-            # only for template specific group users
-            return {record_id: template_instructions[0][0] for record_id in res_ids}
-
         # prepare template variables
         variables = self._render_eval_context()
         if add_context:
@@ -438,7 +430,7 @@ class MailRenderMixin(models.AbstractModel):
 
             try:
                 results[record.id] = render_inline_template(
-                    template_instructions,
+                    parse_inline_template(str(template_txt)),
                     variables
                 )
             except Exception as e:
